File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int solve(int pos, vector<int>& nums, vector<int>& dp){
        if(pos >= nums.size()){
            return 0;
        }
        if(dp[pos] != -1) return dp[pos];
        int left = solve(pos + 1, nums, dp);
        int right = nums[pos] + solve(pos + 2, nums, dp);
        return dp[pos] = max(left, right);
    }
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size() + 1, -1);
        return solve(0, nums, dp);
    }
};
/*
// @Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    int rob(vector<int>& nums) {
        int length = nums.size();
        
        int include = 0;
        int exclude = 0;
        
        for(int i = 0; i < length; i++){
            int in = include, ex = exclude;
            include = ex + nums[i];
            exclude = max(in, ex);
        }
        
        return max(include, exclude);
    }
};


// Time Complexity: O(N), here N = Length of the Given array
// Space Complexity: O(1), because we are using only Constant memory
*/

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # dp = {}
        # def dfs(i):
        #     if i >= len(nums):
        #         return 0
        #     if i in dp:
        #         return dp[i]
        #     rob = nums[i] + dfs(i + 2)
        #     not_rob = dfs(i + 1)
        #     dp[i] = max(rob, not_rob)
        #     return dp[i]
        # return dfs(0)
        prev = 0
        prevprev = 0
        for n in nums:
            best = max(prevprev + n, prev)
            prevprev = prev
            prev = best
        return max(prev, prevprev)